### PR TITLE
chore(gatsby): migrate date to TypeScript

### PR DESCRIPTION
**Pull request by sasurau4**
## Description

Part of #21995 

I migrated `types/date.js` to typescript.

## Related Issues

#21995

### CODE REVIEW

**Comment by LekoArts** [comment]
Sorry for the late response on this. Can you please rebase onto master? Thanks!

File: packages/gatsby/src/schema/types/__tests__/date.js [hunk 1]
```diff
@@ -3,7 +3,8 @@ const { actions } = require(`../../../redux/actions`)
 const { build } = require(`../..`)
 const withResolverContext = require(`../../context`)
 const { defaultResolver } = require(`../../resolvers`)
-const { isDate, looksLikeADate } = require(`../date`)
+import { isDate, looksLikeADate } from "../date"
+require(`../../../db/__tests__/fixtures/ensure-loki`)()
```
Comment (sasurau4) [inline packages/gatsby/src/schema/types/__tests__/date.js:7]: hmm 🤔 
I think I missed the rebase. I look into this.

Comment (pvdz) [reply packages/gatsby/src/schema/types/__tests__/date.js:7]: I think you can drop all changes to this file except for the actual `import` change ;)

File: packages/gatsby/src/schema/infer/inference-metadata.ts [hunk 2]
```diff
@@ -509,7 +509,7 @@ const initialMetadata = (state?: object): ITypeMetadata => {
     total: 0,
     ignoredFields: undefined,
     fieldMap: {},
-    ...state,
+    ...state
```
Comment (pvdz) [inline packages/gatsby/src/schema/infer/inference-metadata.ts:512]: Please drop changes in this file (maybe update your Prettier or figure out why the git commit hooks don't run)

File: packages/gatsby/src/schema/types/date.ts [hunk 3]
```diff
@@ -178,18 +180,25 @@ function looksLikeADate(value) {
  * @param {*} value
  * @return {boolean}
  */
-function isDate(value) {
+export function isDate(value: MomentInput): boolean {
   const momentDate = moment.utc(value, ISO_8601_FORMAT, true)
   return typeof value !== `number` && momentDate.isValid()
 }
 
+interface IFormatDateArgs {
```
Comment (pvdz) [inline packages/gatsby/src/schema/types/date.ts:188]: Can you please put the type declarations at the top, below the imports?

File: packages/gatsby/src/schema/types/date.ts [hunk 4]
```diff
@@ -210,7 +219,26 @@ const formatDate = ({
   return normalizedDate
 }
 
-const getDateResolver = (options = {}, fieldConfig) => {
+interface IDateResolverOption {
```
Comment (pvdz) [inline packages/gatsby/src/schema/types/date.ts:222]: Can you please put the type declarations at the top, below the imports?

File: packages/gatsby/src/schema/types/__tests__/date.js (hunk 1 above)
Comment (pvdz) [reply packages/gatsby/src/schema/types/__tests__/date.js:7]: After formatbot, the require just needs to be dropped (loki was removed). Thanks!

Comment (pvdz) [reply packages/gatsby/src/schema/types/__tests__/date.js:7]: (Rebase bot did not fix this, you'll have to do it)

**Pull request merged**